### PR TITLE
Get rid of `missing docs` in a number of modules

### PR DESCRIPTION
**Pull request by playfulFence**
### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [x] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [x] My changes are in accordance to the [esp-rs API guidelines](https://github.com/esp-rs/esp-hal/blob/main/documentation/API-GUIDELINES.md)

#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/documentation/CONTRIBUTING.md) and followed its instructions.

#### Description
Closes https://github.com/esp-rs/esp-hal/issues/1938

### CODE REVIEW

File: esp-backtrace/src/riscv.rs [hunk 1]
```diff
@@ -12,45 +12,93 @@ pub(super) const RA_OFFSET: usize = 4;
...
 #[doc(hidden)]
-#[allow(missing_docs)]
 #[derive(Default, Clone, Copy)]
 #[cfg_attr(feature = "defmt", derive(defmt::Format))]
 #[repr(C)]
 pub(crate) struct TrapFrame {
+    /// Return address, stores the address to return to after a function call or
+    /// interrupt.
     pub ra: usize,
+    /// Temporary register t0, used for intermediate values.
```
Comment (bugadani) [inline esp-backtrace/src/riscv.rs:22]: This is terrible, can we turn these into arrays?

Comment (playfulFence) [reply esp-backtrace/src/riscv.rs:22]: I agree, but this is not the subject of this PR, so I would not like to make any changes here in terms of this commit tbh.

File: esp-hal/src/ecc.rs [hunk 2]
```diff
@@ -55,30 +53,47 @@ pub enum Error {
...
+    /// Modular addition mode.
     ModAdd                  = 8,
     #[cfg(esp32h2)]
+    /// Modular subtraction mode.
     ModSub                  = 9,
     #[cfg(esp32h2)]
+    /// Modular multiplication mode.
     ModMulti                = 10,
     #[cfg(esp32h2)]
+    /// Modular multiplication mode.
```
Comment (bugadani) [inline esp-hal/src/ecc.rs:96]: This was meant to be division, right?

Comment (playfulFence) [reply esp-hal/src/ecc.rs:96]: Oops, yeah, will be fixed, thanks!

File: esp-hal/src/gpio/mod.rs [hunk 3]
```diff
@@ -170,31 +170,40 @@ pub struct RtcOutput;
...
     I20mA = 2,
+    /// Drive strength of approximately 40mA.
     I40mA = 3,
 }
 
 /// Alternate functions
 #[derive(PartialEq)]
-#[allow(missing_docs)]
 pub enum AlternateFunction {
+    /// Alternate function 0.
```
Comment (bugadani) [inline esp-hal/src/gpio/mod.rs:187]: This is a case where I prefer missing documentation to whatever this is. These doc comments are supposed to be helpful, they shouldn't just echo some implementation detail. Please link to somewhere where the user can get more information about these.

File: esp-hal/src/i2c.rs [hunk 4]
```diff
@@ -1528,6 +1545,7 @@ pub trait Instance: crate::private::Sealed {
     }
 
     #[cfg(not(any(esp32, esp32s2)))]
+    /// Reads all bytes from the RX FIFO
```
Comment (bugadani) [inline esp-hal/src/i2c.rs:1548]: I'm not sure I want to be as pedantic as the LLVM people, but most other comments end with a period, and I guess we should be consistent in that :)

```suggestion
    /// Reads all bytes from the RX FIFO.
```

File: esp-hal/src/i2c.rs [hunk 5]
```diff
@@ -1549,6 +1567,7 @@ pub trait Instance: crate::private::Sealed {
     }
 
     #[cfg(any(esp32, esp32s2))]
+    /// Reads all bytes from the RX FIFO
```
Comment (bugadani) [inline esp-hal/src/i2c.rs:1570]:
```suggestion
    /// Reads all bytes from the RX FIFO.
```

File: esp-hal/src/i2c.rs [hunk 6]
```diff
@@ -1618,6 +1640,8 @@ pub trait Instance: crate::private::Sealed {
 
         Ok(())
     }
+
+    /// Checks for I2C transmission errors and handles them.
```
Comment (bugadani) [inline esp-hal/src/i2c.rs:1644]: You might want to expand on what "handling" means here.

File: esp-hal/src/i2c.rs [hunk 7]
```diff
@@ -1658,6 +1682,7 @@ pub trait Instance: crate::private::Sealed {
         Ok(())
     }
 
+    /// Updates the configuration of the I2C peripheral.
```
Comment (bugadani) [inline esp-hal/src/i2c.rs:1685]: With what?

File: esp-hal/src/lcd_cam/lcd/mod.rs [hunk 8]
```diff
@@ -14,36 +14,53 @@ use crate::{peripheral::PeripheralRef, peripherals::LCD_CAM};
 
 pub mod i8080;
 
+/// Represents an LCD interface peripheral.
```
Comment (bugadani) [inline esp-hal/src/lcd_cam/lcd/mod.rs:17]: You call the camera and i8080 an interface, not an interface peripheral. We might want to be consistent with those.

File: esp-hal/src/lcd_cam/mod.rs [hunk 9]
```diff
@@ -5,7 +5,7 @@
 //! used simultaneously. For more information on these modules, please refer to
 //! the documentation in their respective modules.
 
-#![allow(missing_docs)] // TODO: Remove when able
+// #![allow(missing_docs)] // TODO: Remove when able
```
Comment (bugadani) [inline esp-hal/src/lcd_cam/mod.rs:8]: Remove this please when you're ready :)

File: esp-hal/src/lcd_cam/mod.rs [hunk 10]
```diff
@@ -21,12 +21,16 @@ use crate::{
     InterruptConfigurable,
 };
 
+/// Represents a combined LCD and Camera interface peripheral.
```
Comment (bugadani) [inline esp-hal/src/lcd_cam/mod.rs:24]: Same here

File: esp-hal/src/otg_fs.rs [hunk 11]
```diff
@@ -49,16 +47,20 @@ use crate::{
 };
 
 #[doc(hidden)]
+/// Trait representing the USB D+ (data plus) pin.
```
Comment (bugadani) [inline esp-hal/src/otg_fs.rs:50]: Hold on, these trigger the warning even with `#[doc(hidden)]`? 🤯

File: esp-hal/src/otg_fs.rs [hunk 12]
```diff
@@ -129,7 +131,7 @@ unsafe impl<'d> UsbPeripheral for Usb<'d> {
         80_000_000
     }
 }
-
+/// Asynch functionality
```
Comment (bugadani) [inline esp-hal/src/otg_fs.rs:134]: You're calling this "Async" somewhere else.

File: esp-hal/src/peripheral.rs [hunk 13]
```diff
@@ -355,6 +358,7 @@ mod peripheral_macros {
             $(#[$cfg])?
             #[derive(Debug)]
             #[allow(non_camel_case_types)]
+            /// Macro to create a peripheral structure with a virtual interface.
```
Comment (bugadani) [inline esp-hal/src/peripheral.rs:361]: I don't get this. Isn't the documentation placed on the generated struct, and not the macro?

File: esp-hal/src/rtc_cntl/rtc/esp32s3.rs [hunk 14]
```diff
@@ -39,6 +39,7 @@ pub(crate) fn configure_clock() {
 // System Reset: Reset the whole digital system, including RTC sub-system
 // Chip Reset:   Reset the whole chip, including the analog part
 
+/// SOC reset reason
```
Comment (bugadani) [inline esp-hal/src/rtc_cntl/rtc/esp32s3.rs:42]: This is the third variation of this comment :D `SOC Reset Reason.` vs `SOC Reset Reason` vs `SOC reset reason`. Pick one, please.

File: esp-hal/src/rtc_cntl/sleep/esp32c3.rs [hunk 15]
```diff
@@ -264,16 +302,19 @@ bitfield::bitfield! {
...
     /// Number of rtc_fast_clk cycles to wait for PLL clock to be ready
     pub u8, pll_wait, set_pll_wait: 23, 16;
-    // Perform clock control related initialization.
+    /// Perform clock control related initialization.
     pub clkctl_init, set_clkctl_init: 24;
-    // Perform power control related initialization.
+    /// Perform power control related initialization.
     pub pwrctl_init, set_pwrctl_init: 25;
-    // Force power down RTC_DBOOST
+    /// Force power down RTC_DBOOST
```
Comment (bugadani) [inline esp-hal/src/rtc_cntl/sleep/esp32c3.rs:309]:
```suggestion
    /// Force power down `RTC_DBOOST`.
```

File: esp-hal/src/soc/esp32s3/cpu_control.rs [hunk 16]
```diff
@@ -132,9 +135,11 @@ impl<'a> Drop for AppCoreGuard<'a> {
     }
 }
 
+/// Represents error that can occur in the system.
```
Comment (bugadani) [inline esp-hal/src/soc/esp32s3/cpu_control.rs:138]:
```suggestion
/// Represents errors that can occur in the system.
```

To be fair, "can occur in the system" is VERY vague, especially for the single variant this enum has 🫠

File: esp-hal/src/soc/esp32s3/cpu_control.rs [hunk 17]
```diff
@@ -132,9 +135,11 @@ impl<'a> Drop for AppCoreGuard<'a> {
...
 #[derive(Debug, Clone, Copy, PartialEq)]
 #[cfg_attr(feature = "defmt", derive(defmt::Format))]
 pub enum Error {
+    /// Core is already running.
```
Comment (bugadani) [inline esp-hal/src/soc/esp32s3/cpu_control.rs:142]:
```suggestion
    /// The core is already running.
```

**Comment by bugadani** [comment]
Sorry for the individual comments, I didn't expect to leave this many. I've also stopped commenting on missed periods. In general, I don't necessarily agree with replacing the attribute with pretty trivial docs. I would personally either leave the `allow` for some of these, where listing e.g. priority levels doesn't really add any value.

File: esp-hal/src/otg_fs.rs (hunk 11 above)
Comment (playfulFence) [reply esp-hal/src/otg_fs.rs:50]: Nope, just wanted to leave a comment there as well.

**Comment by playfulFence** [comment]
Thanks a lot for a proper review!

Such documentation is always painful to write, and even worse - to review. I disagreed with some points of the review, and I did not include them in the fixing commit. If there will be more comments on the same things, I will correct them too.

**Review by bugadani — APPROVED** [review]
I continue to not like everything but we can always come back later. Thank you for grinding through these things!

File: esp-hal/src/gpio/mod.rs (hunk 3 above)
Comment (MabezDev) [reply esp-hal/src/gpio/mod.rs:187]: I agree, whilst we still need these mechanical comments to satisfy the missing docs lint, I'd like to see a top level doc on each enum with a bit more explanation.

File: esp-hal/src/hmac.rs [hunk 18]
```diff
@@ -109,6 +118,8 @@ impl<'d> Hmac<'d> {
         }
     }
 
+    /// Releases the HMAC peripheral, returning ownership of the `HMAC`
```
Comment (MabezDev) [inline esp-hal/src/hmac.rs:121]: Careful not to re-introduce this when rebasing, we removed it in #1972

**Review by MabezDev — CHANGES_REQUESTED** [review]
Thanks for working on this! I know docs isn't the most fun job, but it's important.

I agree with some of your push back in the reviews, and have marked those as resolved but the rest, along with my own should be resolved before we can merge this.

File: esp-hal/src/i2c.rs (hunk 6 above)
Comment (playfulFence) [reply esp-hal/src/i2c.rs:1644]: Resolved

File: esp-hal/src/i2c.rs (hunk 7 above)
Comment (playfulFence) [reply esp-hal/src/i2c.rs:1691]: Fixed

File: esp-hal/src/peripheral.rs (hunk 13 above)
Comment (playfulFence) [reply esp-hal/src/peripheral.rs]: Fixed

**Comment by bugadani** [comment]
@MabezDev  can we rip off the bandaid here, please? This will cause me some headaches in #1970 and I'd love to start resolving merge conflicts ASAP.

**Review by MabezDev — APPROVED** [review]
LGTM, thanks!

**Pull request merged**